crates/storage/src/collections/slice/mod.rs

```diff
@@ -0,0 +1,384 @@
+    }
+
+    /// Returns the first and all the rest of the elements of the slice, or None if it is empty.
+    #[inline]
+    pub fn first_mut(&mut self) -> Option<&mut T::Item> {
+        self.get_mut(0)
+    }
+
+    /// Returns the first and all the rest of the elements of the slice, or None if it is empty.
+    #[inline]
+    pub fn split_first(&self) -> Option<(&T::Item, Slice<&T>)> {
+        let first = self.first()?;
+        Some((
+            first,
+            Slice::new(self.range.start + 1..self.range.end, &self.backing_storage),
+        ))
+    }
+
+    /// Returns the first and all the rest of the elements of the slice, or None if it is empty.
+    #[inline]
+    pub fn split_first_mut(&mut self) -> Option<(&mut T::Item, SliceMut<&T>)> {
+        let first =
+            // Safety: we have exclusive access to the slice through the &mut receiver, thus this
+            // mutable borrow is guaranteed to be unique.
+            unsafe {
+            self.backing_storage.get_mut(self.range.start)?
+        };
+        Some((
+            first,
+            // Safety: By taking &mut self, we ensure that other getters of items become cannot be
+            // called until the newly returned SliceMut is dropped. Thus only a single slice has
+            // mutable access to the underlying items.
+            unsafe {
+                SliceMut::new(self.range.start + 1..self.range.end, &self.backing_storage)
+            },
+        ))
+    }
+
+    /// Returns the last and all the rest of the elements of the slice, or None if it is empty.
+    #[inline]
+    pub fn split_last_mut(&mut self) -> Option<(&mut T::Item, SliceMut<&T>)> {
+        let last =
+        // Safety: we have exclusive access to the slice through the &mut receiver, thus this
+        // mutable borrow is guaranteed to be unique.
+        unsafe {
+            self.backing_storage.get_mut(self.range.end)?
+        };
+        Some((
+            last,
+            // Safety: By taking &mut self, we ensure that other getters of items become cannot be
+            // called until the newly returned SliceMut is dropped. Thus only a single slice has
+            // mutable access to the underlying items.
+            unsafe {
+                SliceMut::new(self.range.start..self.range.end - 1, &self.backing_storage)
+            },
+        ))
+    }
+
+    /// Returns the last and all the rest of the elements of the slice, or None if it is empty.
+    #[inline]
+    pub fn split_last(&self) -> Option<(&T::Item, Slice<&T>)> {
+        let first = self.last()?;
+        Some((
+            first,
+            Slice::new(self.range.start..self.range.end - 1, &self.backing_storage),
+        ))
+    }
+
+    /// Creates a new `SliceMut`.
+    ///
+    /// # Safety:
+    ///
+    /// The caller must ensure that mutable slices do not overlap.
+    #[inline]
+    pub(crate) unsafe fn new(range: Range<u32>, backing_storage: T) -> SliceMut<T> {
```

Comment:
Same nitpick about the location. 

Also, why'd this `pub(crate)` instead of just `pub`?

Reply:
That should indeed be just `pub`, unless we do not want library authors to be able to create slices.